### ISSUE TITLE
Parser still confusing opcode and variable names

### ISSUE THREAD

**Issue by vlazzarini** [csound/csound#1964]
The new parser is supposed to tell variable names from opcodes, but there are some issues still.

For example, let's say we have a variable declared as `fmax:i`, which is also the name of an opcode. If we declare it by assigning a value,

```
fmax:i = p5
```

it works as expected. However, if we use `init` as the means of a first declaration,

```
fmax:i init p5
```

then it doesn't:

```
error:  Unable to find opcode entry for 'fmax' with matching argument types:

Found:
  (null) fmax S

Candidates:
  i[] fmax i[]i[]
  k[] fmax k[]k[]
  i[] fmax i[]i
  k[] fmax k[]k
  i[] fmax i[]i[]
  k[] fmax k[]k[]
  i[] fmax i[]i
  k[] fmax k[]k
```

It is also curious that it finds `(null) fmax S`, when there is no S type in the expression at all (p5, another variable or a literal yield the same error message).

If the variable name is changed in such a way that an opcode with that name does not exist, the code compiles correctly.

**Comment by kunstmusik** [comment]
Interesting, I think this is probably a byproduct of the disambiguation process trying to function more like earlier parsers where opcode names took precedence. Might be possible to just change that precedence and consult var table first. Will need to research further to see if doing that will remain backwards compatible.

**Comment by vlazzarini** [comment]
@scottsitar maybe we could use your flex/bison expertise to tackle this one?
I checked today and this is still failing to parse.

**Comment by scottsitar** [comment]
@vlazzarini Sure, I can look into it.

**Comment by vlazzarini** [comment]
thanks.

**Comment by scottsitar** [comment]
From an initial look, I think the issue might not be flex/bison building the parse tree, but rather the semantic analysis afterwards.  For the following simple instrument:

```
instr 1
    fmax:i init p4
endin
```

I get this parse tree, which seems reasonable to me:

```
<tree addresses="(0x600003120310 : 0x0)" type="288" name="INSTR_TOKEN" loc="7:/Users/audiodev/issues/1964/test.csd">
 <treeleft addresses="(0x600003120130 : 0x6000000205d0)" type="299" name="INTEGER_TOKEN" value="1" loc="6:/Users/audiodev/issues/1964/test.csd">
 </treeleft>
 <treeright addresses="(0x6000031202b0 : 0x0)" type="284" name="T_OPCALL" varname="init" loc="8:/Users/audiodev/issues/1964/test.csd">
  <treeleft addresses="(0x6000031201f0 : 0x600000020650)" type="284" name="T_OPCALL" varname="fmax" loc="7:/Users/audiodev/issues/1964/test.csd">
   <treeleft addresses="(0x600003120190 : 0x600000020610)" type="294" name="T_IDENT" varname="fmax" loc="7:/Users/audiodev/issues/1964/test.csd">
   </treeleft>
   <treeright addresses="(0x600003120250 : 0x600000020690)" type="294" name="T_IDENT" varname="p4" loc="7:/Users/audiodev/issues/1964/test.csd">
   </treeright>
  </treeleft>
 </treeright>
</tree>
```

But then something happens in csound_orc_semantics, verify_opcode which I don't fully understand yet.  What it looks like is that this might be causing problems:

https://github.com/csound/csound/blob/develop/Engine/csound_orc_semantics.c#L1132

During semantic analysis, `get_arg_string_from_tree` is getting called on the following tree:

```
<tree addresses="(0x6000031201f0 : 0x600000020650)" type="285" name="T_FUNCTION" fname="init" loc="7:/Users/audiodev/issues/1964/test.csd">
 <treeright addresses="(0x600003120250 : 0x600000020690)" type="294" name="T_IDENT" varname="p4" loc="7:/Users/audiodev/issues/1964/test.csd">
 </treeright>
</tree>
```

This calls `get_arg_type2` which tries to find the OENTRY for `init p4`.  The candidates listed in order are:

```
OENTRY opname: init.C
OENTRY opname: init.C
OENTRY opname: init.S
OENTRY opname: init.Si
OENTRY opname: init
OENTRY opname: init.i
OENTRY opname: init.k
OENTRY opname: init.a
OENTRY opname: init.instr
OENTRY opname: init.f
OENTRY opname: init.i
OENTRY opname: init.k
OENTRY opname: init.a
OENTRY opname: init.S
OENTRY opname: init.0
```

The input arg string is 'p' (seems fine).  And based on the argument type matching, it's deciding that `init.Si` is the (first) match.

I don't understand the opcode matching/argument type code well enough to give any good explanations or proposals for a fix, but I'll keep looking to understand this some more.

Does this spur any ideas for folks that know the code better?

**Comment by scottsitar** [comment]
Actually, thinking about this some more, is the parse tree I originally posted as reasonable as I thought?  Is it actually saying that it parsed as (in functional notation):

```
fmax:i(init(p4))
```

That would explain the sequence of calls I was seeing:

- find OENTRY for `init(p4)` => `init.Si`, return type of S
- find OENTRY for `fmax:i(S)` => not found => error

In general when disambiguating statements like this, do we try to deterministically figure out the best candidate, try that, and then fail if it doesn't work?  Or do we rank all possible candidates in some order, and try them from most to least likely until a match is found?

**Comment by vlazzarini** [comment]
That doesn't look right as we should have produced

```
fmax:i init p4
```

with `init.i` as the opcode

if you change `fmax` to  say `fpax`, you'll get that

```
<ast>
<tree addresses="(0x106e07490 : 0x0)" type="288" name="INSTR_TOKEN" loc="9:teststr.csd">
 <treeleft addresses="(0x106e07800 : 0x107400c10)" type="299" name="INTEGER_TOKEN" value="1" loc="8:teststr.csd">
 </treeleft>
 <treeright addresses="(0x106e076a0 : 0x107400b30)" type="284" name="T_OPCALL" varname="fpax" loc="9:teststr.csd">
  <treeleft addresses="(0x106e07750 : 0x107400ba0)" type="294" name="T_IDENT" varname="fpax" loc="9:teststr.csd">
  </treeleft>
  <treeright addresses="(0x106e075f0 : 0x107400ac0)" type="294" name="T_IDENT" varname="p4" loc="9:teststr.csd">
  </treeright>
 </treeright>
</tree>
</ast>
```

resulting in

```
instrument 1 of 200: 0x107602360 
init.i args:	fpax	p4
```

**Comment by scottsitar** [comment]
Ah, thanks for the confirmation!

Yeah, I don't yet understand why it was matching `init.Si` when the input argtype was `p`.  During parse tree construction, it does seem to be reducing via the right rule:

```
opcall : out_arg_list identifier expr_list NEWLINE
```

which is a good sign.  I'll keep experimenting with it.

**Comment by vlazzarini** [comment]
Given that a token that does not match an opcode name is parsed OK, I think your initial hunch that the problem may be in disambiguation is probably right.

At some point, there is a decision that resolves fmax:i as an opcode instead of an argument, which gets it wrong. When fpax is used, it doesn't find an opcode match and it goes into a different direction.

**Comment by scottsitar** [comment]
I've got a quick question about T_OPCODE trees.  I've been studying what `convert_statement_to_opcall` does under various conditions, and it seems to me that the shape that T_OPCODE trees can take depends on the stage of disambiguation they've gone through.  In particular, some special cases that get used for recognizing certain cases are:

- when the T_OPCODE tree itself doesn't have a corresponding ORCTOKEN (so, the value field is null)
- when there is a corresponding ORCTOKEN (seems to be already some processing done on this tree)
- when the left tree is also an T_OPCODE and the right tree is null (seems to be the initial state before any disambiguation)

Is this far off base?  Before I started reading this carefully, I was working under the assumption that a T_OPCODE tree would always have the out args as the left tree and the in args as the right tree, but this seems like it might only be once all the transformations have happened?

Also, with a view towards a potential fix, the comment above `convert_statement_to_opcall` mentions that it now needs the type table to make its decisions.  However, it seems that this parameter isn't used anywhere in the function.  Is the comment a statement of intent, and it just hasn't been implemented yet?  I suspect it might want to go around the `ksubst init -1 / TODO` comments in the "`left == T_OPCALL, right == NULL`" branch?

**Comment by vlazzarini** [comment]
@kunstmusik wrote this code, so he will have more details but as far as I can see, the reason for having the typetable is that the code could look at the currently allocated variables in the different pools and see if the variable exists there. From a first look, the fact that the typetable arg is not used is in contradiction with the comment above it, so it may either be incomplete or a late design change.

However, for the present case, I'd say this:

- `convert_statement_to_opcall()` seems to be called before `add_args()`, 
- `fmax` as an output argument is added to the variable pool in `add_arg()` called by `add_args()`
- therefore even if the typetable is used in the search, I don't think it would help this case, since `fmax` would not be there.

It may help however the parsing of

```
fmax:i = 0
print fmax
```

which is failing. Note that `fmax:i = 0` on its own parses.

**Comment by vlazzarini** [comment]
@scottsitar I have the solution for this problem now. Here's my thinking:

- opcodes are symbols in the system/language
- if we go by an ODR (which we do for all variables etc), they should not be redefined, so `fmax:i = p5` is a synterr.
- we have to deal with this correctly rather than give a confusing error message as we do now.

So it happens that a side effect of the work I am doing on a branch (https://github.com/csound/csound/tree/feature/opcodetype),  any assignment etc is dealing with this nicely:

```
error:  fmax:i type mismatch for variable fmax:OpcodeDef
```

So I think with this approach we can move forward.

EDIT: I just realised the issue was not assignment but 'init' and that still is causing a problem. I don't think my approach fixes that, so ignore what I said. I am still experimenting with things in that branch. ODR is probably not applicable here alright.

**Comment by joachimheintz** [comment]
one related question here as i am on it in another context: is it 
supposed to work that a variable name can be an opcode name like

   tone:a = poscil(ramp,freq) ?

or is this a clear restriction / exclusion?

**Comment by vlazzarini** [comment]
> one related question here as i am on it in another context: is it 
> supposed to work that a variable name can be an opcode name like
> 
>    tone:a = poscil(ramp,freq) ?
> 
> or is this a clear restriction / exclusion?
> 
> 


currently in 7.0 it works like that (see my initial post), but not always (hence the bug report).

**Comment by joachimheintz** [comment]
thanks for the confirmation.  i saw that the line

tone:a = poscil(.2,400)

gets this error message in curent develop:

error:  Unable to find opcode entry for 'tone' with matching argument types:

(but works with "stone" instead od "tone")

**Comment by vlazzarini** [comment]
yes, that's the bug. We may not be able to allow that, I don't know if we can, but at least the error message should be improved. At the moment, the parser just gets confused.

**Comment by scottsitar** [comment]
Oh interesting.  I've been playing around with the following heuristic for disambiguation.  For a line of the form:

```
A   B   C, D, E
```

this goes through `convert_statement_to_opcall` with a special shape of parse tree.  The upside is that (it seems) we can always tell when we're in this case and easily pull out A and B.  So, my idea is:

- if B is not in any pool in the type table, then parsing this line with A as an opcode will never work (arguments must be in the type table by the time the opcode statement is verified).  Assume B is an opcode and proceed to reshape the tree.
- if B is in the type table, then we need to respect that and assume that A is an opcode

Have I made any incorrect assumptions?  I've got the original statement from the bug report parsing correctly now, but I haven't 100% convinced myself that this won't lead to other confusing issues.  Unfortunately, the error messages are not always better yet.

**Comment by vlazzarini** [comment]
I think the assumption is correct. A and B cannot be variables at the same time, so if B is a variable, A is an opcode, and vice-versa. They can be two opcodes if B is inline (e.g. B() etc) like in the bug you fixed.

Btw, one thing I noticed when doing the work in the branch was this comment

https://github.com/csound/csound/blob/2303e5b486df0bac4ac88e332a783c740307811a/Engine/csound_orc_semantics.c#L1809

does it have any bearing on the issue? I stuck a check there in the branch but I don't think it did anything for this.

**Comment by scottsitar** [comment]
Yes, that's definitely the block I'm sniffing around in, since I believe that's the shape of the tree that identifies this particular case.  I'm still working on getting the error messages to be better since if we let it get to verify_opcode with an incorrect tree, it can print out some misleading advice.

**Comment by scottsitar** [comment]
I think I've been barking up the wrong tree (pun intended?) this whole time.  `convert_statement_to_opcall` will get called twice in succession, once on the first special tree generated from the opcall bison rule, and once on the "first pass simplified" version (essentially just the left subtree but keeping the root lexeme).  I've been focusing on the first call as the one that's getting it wrong, but it looks like it's actually the second call that's doing the incorrect transformation.  I'm now pretty convinced that it's this block of code:

https://github.com/csound/csound/blob/develop/Engine/csound_orc_semantics.c#L1794

which is taking a line of the form `fmax:i init p4`, and immediately digging into the left subtree to see if it's an opcode, without ever considering the root lexeme - which it knows it has at this point!  The comment just above makes it look like it's expecting a statement of the form `xout exp(0)` at this point, and the transformation it does will be valid for something like that.  It will mangle a line like we've got here, though.  I think we just need a stricter check in the conditions on that if statement.  For a quick test, if I just comment it all out, the example compiles correctly.

Sorry for being a little pokey on this one.  I'm still internalizing all of the different parser structures, but the fog is starting to clear a little bit on it now.  I need to do a little bit more experimentation on what functions and opcodes will look like in tree form at this point in the process to verify a few assumptions I'm making, but this might actually turn out to be a small change once everything is said and done.

**Comment by vlazzarini** [comment]
interesting, thanks for the report!

**Comment by joachimheintz** [comment]
this detective work is quite fascinating for me as spectator, but must 
make headache for the ones diving into this code ...

**Comment by scottsitar** [comment]
I've submitted a PR for this (https://github.com/csound/csound/pull/2090), although I think it might be good for someone with a larger collection of test scripts to give it a go to make sure it doesn't break anything.  Apart from the example above in the problem report (which the original code wasn't intended to handle it seems), I wasn't able to construct an example that could hit that block of code.  Also, after the fix to detect opcodes with parenthesized first arguments vs function starts, it seems like the intended case is correctly handled without this transformation now.

As a test, I just removed that case and ran the tests, and everything passed.  So, I've added a test case to make sure that the issue above remains fixed, but if anyone could point me in the direction of an example where the old code is necessary, I think that would also help to test this change.  If removing this looks fine to people, then it should be ready for review.

**Comment by scottsitar** [comment]
Oof, sorry, I spoke too soon.  With this change, the `fmax:i init p4` parses correctly, but if you put a `print fmax:i` right after it, it will now fail to parse.  The plot thickens...

**Comment by scottsitar** [comment]
Okay, I've put up another PR which should fix both problems: https://github.com/csound/csound/pull/2091.

Apologies for the noise.  The same caveats as above apply to this new change - if anyone has a larger set of scripts they can run this against to make sure removing this one transformation is alright, that would be good.  The second problem I found turned out to be unrelated to the initial issue, but I thought I'd be a good citizen and try to fix both.
